File: A11 Gernal problems/Leetcode 8 StingToInteger/String to Integer.java
/*
Atoi ==> Ascii to Integer

Step 1: first remove the white space at the both end of string; 
Step 2: corner case 
Step 3: check if the first character is + or -; if neither, don't need to idx ++; it is still at first index
step 4: iteration 
    validate pre-character and /or post-character codition
                if(!Character.isDigit(s.charAt(i))){
                return (int) sum * sign; 
            }
    otherwise, it is character, so read it
            sum = sum * 10 + s.charAt(i) - '0';
    check the range of the digital
*/ 


class Solution {
    public int myAtoi(String s) {
        s = s.trim();    
        if(s == null || s.length() ==0) return 0;
        
        int idx =0;
        int sign = 1;
        char c = s.charAt(0);
        if (c =='+'){
            sign =1; 
            idx ++;
        }else if (c == '-'){
            sign = -1; 
            idx  ++;
        }
        
        
        long sum = 0; 
        for (int i = idx; i < s.length(); i++){
            if(!Character.isDigit(s.charAt(i))){
                return (int) sum * sign; 
            }
            sum = sum * 10 + s.charAt(i) - '0';
            
            if (sum > Integer.MAX_VALUE && sign == 1){
                return Integer.MAX_VALUE;
            }
            if ((-1)* sum < Integer.MIN_VALUE && sign == -1){
                return Integer.MIN_VALUE;
            }
        }
        return (int)(sum * sign);
        
    }
}

File: A11 Gernal problems/Leetcode 8 StingToInteger/StringToInteger.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()
        if len(s) == 0 or (len(s) > 0 and not (s[0] in ['+','-'] or s[0].isdigit())):
            return 0

        value = 0
        sign = -1 if s[0] == '-' else 1
        i = 1 if not s[0].isdigit() else 0

        while i < len(s) and s[i].isdigit():
            value = (value * 10) + (ord(s[i]) - ord('0'))
            i += 1
        value *= sign

        if value < -(2**31):
            return -(2**31)
        elif value > (2**31)-1:
            return (2**31)-1

        return value        

##########
Time complexity: O(n)
Space complexity: O(1)